// ipeweb-core-styles/dist/styles/one-dark-colors.style.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var csx_1 = require("csx");
var csx_2 = require("csx");
exports.ColorHelper = csx_2.ColorHelper;
exports.colors = {
    foreground: csx_1.color('#abb2bf'),
    foregroundBold: csx_1.color('#abb2bf'),
    background: csx_1.color('#282c34'),
    darkBackground: csx_1.rgb(31, 33, 40),
    shadow: csx_1.rgba(0, 0, 0, 0.2),
    input: csx_1.rgb(22, 23, 26),
    // Black
    blackDark: csx_1.color('#282c34'),
    blackLight: csx_1.color('#5c6370'),
    // Red
    redLight: csx_1.color('#e06c75'),
    redDark: csx_1.color('#be5046'),
    // Green
    greenLight: csx_1.color('#98c379'),
    greenDark: csx_1.color('#7a9f60'),
    // Yellow
    yellowLight: csx_1.color('#e5c07b'),
    yellowDark: csx_1.color('#d19a66'),
    // Blue
    blueLight: csx_1.color('#61afef'),
    blueDark: csx_1.color('#3b84c0'),
    // Magenta
    magentaLight: csx_1.color('#c678dd'),
    magentaDark: csx_1.color('#9a52af'),
    // Cyan
    cyanLight: csx_1.color('#56b6c2'),
    cyanDark: csx_1.color('#3c909b'),
    // White
    whiteLight: csx_1.color('#abb2bf'),
    whiteDark: csx_1.color('#828997'),
    // Orange
    orangeLight: csx_1.rgb(239, 84, 8),
    orangeDark: csx_1.rgb(134, 49, 3),
};
